### PR TITLE
submission: Int6 MLP3x + Late-K Passthrough + SlidingWindow (val_bpb: 1.1605)

### PR DESCRIPTION
**Pull request by takhir-iota**
# Int6 MLP3x + Late-K Passthrough + SlidingWindow

## Summary

This PR is a 10-minute submission for leaderboard placement, not a record claim.

The submitted run is the best under-cap seed on this lane:

- `seed2025`
- `final_sliding_window_eval_exact stride:64 val_loss:1.95946000 val_bpb:1.16050360`
- `Total submission size quant+zstd: 15,844,924`

The lane stacks four practical improvements on the strong 9-layer, 512-dim GPT recipe:

1. **Int6 mixed quantization + zstd**: `.mlp.`, `.attn.c_q.`, `.attn.c_v.`, and `.attn.proj.` are stored in int6, then compressed with `zstd`.
2. **3x MLP expansion**: `MLP_MULT=3` keeps the wider hidden layer that materially improves score within the byte budget.
3. **Selective K preservation**: `blocks.7.attn.c_k.weight` and `blocks.8.attn.c_k.weight` stay in fp16, while the remaining `c_k` matrices use grouped int8 with `group_size=64`.
4. **Sliding-window evaluation**: `EVAL_STRIDE=64` gives near-full context at evaluation time and is the main improvement over the stride-256 variant.

## Configuration

```text
VOCAB_SIZE=1024 NUM_LAYERS=9 MODEL_DIM=512 NUM_HEADS=8 NUM_KV_HEADS=4
MLP_MULT=3 TIE_EMBEDDINGS=1
MATRIX_LR=0.02 SCALAR_LR=0.02 TIED_EMBED_LR=0.03
MUON_MOMENTUM=0.99 MUON_MOMENTUM_WARMUP_START=0.92 MUON_MOMENTUM_WARMUP_STEPS=1500
WARMDOWN_ITERS=3000 QK_GAIN_INIT=1.7
TRAIN_BATCH_TOKENS=524288 TRAIN_SEQ_LEN=1024
LOWBIT_BITS=6 LOWBIT_STE=0
LOWBIT_NAME_PATTERNS=.mlp.,.attn.c_q.,.attn.c_v.,.attn.proj.
INT8_KEEP_FLOAT_NAME_PATTERNS=tok_emb.weight,blocks.7.attn.c_k.weight,blocks.8.attn.c_k.weight
INT8_GROUP_OVERRIDES=.attn.c_k.:64
SERIAL_COMPRESSOR=zstd
EVAL_STRIDE=64
MAX_WALLCLOCK_SECONDS=600
```

## Command

```bash
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

## Key Metrics

- Training stopped at step **12791/20000** due to the 600s wallclock cap
- Average training step time: **46.91ms**
- Model params: **21,778,504**
- Pre-quant eval: `val_loss:2.0047 val_bpb:1.1873`
- Quantized roundtrip: `val_loss:2.01675174 val_bpb:1.19443398`
- Sliding window (stride=64): `val_loss:1.95946000 val_bpb:1.16050360`
- Sliding-window eval time: **70834ms**
- Code size: **37,988 bytes**
- Total submission size: **15,844,924 bytes**

## Quantization Strategy

The main serializer choice is to spend the remaining bytes on the attention `K` path rather than on broader fp16 promotion:

- `tok_emb.weight`: fp16 passthrough
- `blocks.7.attn.c_k.weight` and `blocks.8.attn.c_k.weight`: fp16 passthrough
- remaining `.attn.c_k.` matrices: grouped int8, `group_size=64`
- `.mlp.`, `.attn.c_q.`, `.attn.c_v.`, `.attn.proj.`: int6
- compressor: `zstd`

This keeps the artifact under `16,000,000` bytes while preserving the highest-value late-layer key projections.

## Additional Seeds

The same lane was rerun on two more under-cap seeds:

- `seed42`: `val_loss:1.96035715 val_bpb:1.16103494`, `15,802,877` bytes
- `seed4242`: `val_loss:1.96595032 val_bpb:1.16434753`, `15,822,568` bytes

Maintainers can place this submission according to the project’s current eligibility and ranking criteria.

### CODE REVIEW

**Comment by takhir-iota** [comment]
The submitted code snapshot is minified because code size counts toward the 16,000,000-byte submission limit. I kept the PR body and submission README detailed so the method is still reviewable, and the included logs/config describe the exact run behavior.

**Comment by MatoTeziTanka** [comment]
### Community Review — submission: Int6 MLP3x + Late-K Passthrough + SlidingWindow (val_bpb: 1.1605)

**BPB:** 1.1605 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `3131670eaa55`, file `records/track_10min_16mb/2026-03-19_Seed2025_Top2K_Stride64/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.05s, dim=512, layers=9, vocab=1024, code=37988 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.05s, dim=512, layers=9, vocab=1024, code=37988 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
